Scripted metric aggregations: add deprecation warning and system property to control legacy params

rationull opened the pull request:
This is a followup to PR #30111 implementing issue #29328.

This implements the suggestion from @rjernst that the legacy `params._agg/_aggs` should be available by default and disabled via a system property. A deprecation warning is emitted when scripted metric aggregations are used with the deprecated params enabled.

This is to be followed up with another PR removing the deprecated behavior (and updating relevant tests). I have that implemented but will finalize it after making any requested changes here, and presumably it wouldn't get merged until this is ported to the appropriate release branch and a target branch is selected for removing the old params.

elasticmachine commented:
Pinging @elastic/es-search-aggs

colings86 commented:
@rationull @rjernst I think I've missed something here. Was it intended that https://github.com/elastic/elasticsearch/pull/30111 was backported to 6.x so we can deprecate the use of `params._agg/_aggs` in 6.x and remove it in 7.0? If so we need to backport https://github.com/elastic/elasticsearch/pull/30111 to the 6.x branch as its currently only in master

colings86 commented:
@rationull you are right that this PR would be needed to deprecate the old functionality in master even if we don't back-port so this PR is definitely needed. The question is whether #30111 should have been back-ported to 6.x.

@rjernst I am on vacation after today so if I should have back-ported #30111 then I am happy either for you to do the back-port of the commit or I can do it when I return. If you could leave a comment either here or on #30111 indicating what should be done that would be great.

rjernst commented:
Yes, #30111 should be backported to 6.x I will do that next week.

rationull commented:
Thanks @rjernst. Then this should be tagged v6.4.0 too presumably?

rjernst commented:
Yes, I've added the 6.4.0 label.

colings86 commented:
Thanks for backporting this @rjernst

rationull commented:
Updated with latest from master. @colings86 does this look OK to you and would you mind kicking it to Jenkins?

colings86 commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregatorFactory.java:
```diff
@@ -83,10 +83,17 @@ public Aggregator createInternal(Aggregator parent, boolean collectsFromSingleBu
-            aggParams.put("_agg", new HashMap<String, Object>());
+        Object aggState = new HashMap<String, Object>();
+        if (ScriptedMetricAggContexts.deprecatedAggParamEnabled()) {
+            if (!aggParams.containsKey("_agg")) {
```
nit: we tend to prefer `aggParams.containsKey("_agg") == false` as its more obvious to read since `!` can easily be missed

colings86 commented:
@elasticmachine please test this

colings86 commented:
@elasticmachine test this please

rationull commented:
@colings86 thanks, I will address the comment when I address the test failure. Looks like a trivial failure -- I neglected to add the expected warning headers to the docs.

rationull commented:
@colings86 fixed the style nit and the test failures. The docs sample test setup is pretty slick!

Would you mind kicking this to CI again?

colings86 commented:
@elasticmachine test this please

colings86 commented:
@rationull the latest failure looks unrelated to your changes but would you be able to update your branch with the latest master and I'll kick off CI again to see if we can get a clean run?

rationull commented:
I agree that this failure looks unrelated. I haven't managed to reproduce it in several runs; I wonder if it's a timing issue?

The `master` build appears to be broken now due to a conflict between c662565f and 305bfea9 so I will check back and update my branch when that is fixed.

rationull commented:
@colings86 the branch is updated now.

colings86 commented:
@elasticmachine test this please

colings86 commented on docs/reference/aggregations/metrics/scripted-metric-aggregation.asciidoc:
```diff
@@ -26,6 +26,7 @@ POST ledger/_search?size=0
 --------------------------------------------------
 // CONSOLE
 // TEST[setup:ledger]
+// TEST[warning:params._agg/_aggs for scripted metric aggregations are deprecated, use state/states (not in params) instead. Set system property es.aggregations.enableDeprecatedScriptedMetricAggParam = false to disable this deprecated behavior.]
```
sorry for not seeing this earlier: I think we should make the test cluster set the system property so the deprecated behaviour is not used?

rjernst replied:
Yes it should be set to the new behavior in all tests, and a specific test added for the bwc behavior. See https://github.com/elastic/elasticsearch/pull/30975/files#diff-407c748a710cb1e33230c1fd06da7f16R694 (although adding the sysprop in BuildPlugin should really have a comment indicating the purpose and when it can be removed).

rationull replied:
Excellent -- thanks for the feedback. I already have a commit in another branch updating the various tests to use the new context variables (PR to follow this one) so I can bring over those changes and add another dedicated test class for the old behavior similar to `ScriptDocValuesMissingV6BehaviourTests` from that example.

I should be able to get back to this some evening this week. In the meantime to make sure we're on the same page, please let me know if this is different from what you had in mind:
- The deprecated behavior will be disabled in general via setting `enableDeprecatedScriptedMetricAggParam` to false centrally in BuildPlugin.groovy. With a comment that this gets remove presumably in 7.0.
- The deprecated behavior will be enabled again for a single dedicated integration test class (because this will get us better coverage across the affected classes vs a unit test).


Some of the test changes are substantial (ScriptedMetricIT specifically) so heads up this review will get bigger.

rationull commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricIT.java:
```diff
@@ -90,42 +91,57 @@
 
-            scripts.put("_agg.add(1)", vars ->
-                    aggScript(vars, agg -> ((List) agg).add(1)));
+            scripts.put("state.list.add(1)", vars ->
```
Some of these scripts and test cases had to be revised to add a `list` var under the agg state since `state` itself can't be custom initialized with an arbitrary type like `params._agg/_aggs` could.

rationull commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricIT.java:
```diff
@@ -167,79 +200,38 @@
                 return newAggregation;
             });
 
-            scripts.put("state.items = new ArrayList()", vars ->
```
The test case that referenced these scripts is removed because it only existed to exercise `state` while the rest of the cases still use `params._agg`.

rationull commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricIT.java:
```diff
@@ -365,52 +358,12 @@ public void testMap() {
         assertThat(numShardsRun, greaterThan(0));
     }
 
-    public void testExplicitAggParam() {
```
This case is not meaningful anymore since `state` is always implicit on the context.

rationull commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricIT.java:
```diff
@@ -365,52 +358,12 @@ public void testMap() {
-        assertThat(totalCount, equalTo(numDocs));
-    }
-
-    public void testMapWithParamsAndImplicitAggMap() {
```
Renamed since `state` is always implicit on the context.

rationull commented on server/build.gradle:
```diff
@@ -334,7 +334,22 @@ if (isEclipse == false || project.path == ":server-tests") {
   }
+
+  // TODO: remove this task dedicated to ScriptedMetricAggParamsV6CompatIT for 7.0
+  task scriptedMetricAggDeprecatedParamIntegTest(type: RandomizedTestingTask,
```
I considered centralizing this similar to the `additionalTest` extension for unit tests but opted not to at this point since it seems like overkill with no other known uses for it. Even `additionalTest` only appears to be used in two spots right now.

rjernst replied:
I don't think we need an IT. The purpose of the test only needs to be that the sysprop is read appropriately. The behavior can and should all be tested in a unit test.

rationull commented:
@colings86 @rjernst added changes based on feedback. The deprecated behavior is now disabled in tests including the docs tests and the allowed warning asserts are removed accordingly. I copied the relevant older content from `ScriptedMetricIT` and made that into the backwards compatibility integration test which is run with the deprecated behavior enabled.

In an effort to deal with one change at a time I haven't merged in `master` again yet -- a CI run when convenient would be appreciated and I'll update the branch again once these latest changes are approved or revised.

rjernst commented on docs/reference/migration/migrate_7_0/aggregations.asciidoc:
```diff
@@ -22,4 +22,5 @@ The object used to share aggregation state between the scripts in a Scripted Met
 being provided via the `params` object as `params._agg`.
 
-The old `params._agg` variable is still available as well.
+The old `params._agg` variable is still available as well. This can be disabled by setting
```
Since this will be backported to 6.x, with the intention of removing in 7.0, I don't think this migration note makes sense. The migration note would be that params._agg is gone, and would be in the followup which removes support in master only.

rationull replied:
Good catch, thanks. I overthought this and was going to remove it in my followup, but you're right; there's no need for this note here now that the target version is more solid.

rationull replied:
I'll revert the doc change such that this file is left as-is until the next PR.

rationull commented:
@rjernst I replaced the bwc integration test with a couple of unit tests based on the existing test base classes for the affected components. I think this should address the review feedback.

Also updated to latest master.

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java:
```diff
@@ -31,6 +33,25 @@
+
+    public static boolean deprecatedAggParamEnabled() {
+        boolean enabled = Boolean.parseBoolean(
+            System.getProperty("es.aggregations.enableDeprecatedScriptedMetricAggParam", "true"));
```
This should be named with underscores. I think the name doesn't need the "deprecated" in it.

rationull replied:
So `enable_scripted_metric_agg_param` or is that too generic?

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java:
```diff
@@ -31,6 +33,25 @@
+            System.getProperty("es.aggregations.enableDeprecatedScriptedMetricAggParam", "true"));
+
+        if (enabled) {
+            DEPRECATION_LOGGER.deprecatedAndMaybeLog("enableDeprecatedScriptedMetricAggParams", AGG_PARAM_DEPRECATION_WARNING);
```
Unfortunately I've lead you in an incorrect direction. After discussing with others, we want these to only be triggered when the deprecated behavior is used. See #31441 for and example doing this in ScriptDocValues for an idea of what must be done within the aggs code (although I think it will be simpler here, since you can do this check and deprecation outside of script invocation, and thus no need for a doPriv block).

rationull commented:
@rjernst I renamed the sysprop per suggestion and merged with latest from master. Awaiting feedback on the deprecation warning per discussion before making further changes.

rationull commented:
@rjernst I've pushed [another branch to my fork](/rationull/elasticsearch/tree/rationull/metric-agg-context-deprecation-and-disable-with-modification-detection) that emits the deprecation warning in the following cases instead of whenever scripted metric aggregations are used:

* An `_agg` param is explicitly provided to the aggregation.
* `params._agg` is NOT modified in the init, map, or combine scripts.
* The `states` context variable is NOT accessed in the reduce script.

It would be possible to write a scripted metric aggregation with scripts that reference `_agg`/`_aggs` and do not trip these checks, although I struggle to think of a such a case that would be a meaningful or likely use of the feature (particularly with respect to the reduce case).

The main benefit of this would be that the deprecation warning won't be emitted when using scripted metric aggs only with the new context variables as directed by the modified docs. If it is more appropriate having the deprecation warning only emitted in these cases then I will merge those changes to this branch and include them in this PR.

rjernst commented on server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java:
```diff
@@ -31,6 +33,25 @@
+    // Public for access from tests
+    public static final String AGG_PARAM_DEPRECATION_WARNING =
+        "params._agg/_aggs for scripted metric aggregations are deprecated, use state/states (not in params) instead. " +
+        "Set system property es.aggregations.enable_scripted_metric_agg_param = false to disable this deprecated behavior.";
```
We typically give the actual command line, something like:

```
-Des.aggregations.enable_scripted_metric_agg_param = false to jvm.options to disable...
```

rationull replied:
Updated based on the format of the joda time ScriptDocValues deprecation message that's also currently active.

colings86 commented:
@rationull I think I'm in a very similar position to @rjernst here. I think your alternative is a really good idea and I am also a bit torn between the two option. Ultimately though I think I'm inclined to go with this existing change because it makes sure that every use of the scripted_metric aggregation is definitely covered and will warn the user of the upcoming breaking change. I would like to ensure that we don't have any edge cases where the user is surprised by the breaking change because we didn't log a deprecation warning.

rationull commented:
If there's no clear winner then I think using the simpler and less nuanced approach is a good choice. Nothing intrinsically wrong with a deprecation warning related to a feature you're using even if it doesn't require any changes, but lack of a deprecation warning in a feature that then burns you during a planned upgrade would be frustrating.

The `ctx_in_params` deprecation that was just added looks like it uses this strategy as well, emitting the warning when the field is added rather than trying to detect use somehow.

Regardless I am glad to have implemented both alternatives because it was a good exercise thinking through the various cases and how we can get at more direct knowledge of usage of this kind of feature. Specifically with an eye toward future improvements, it occurred to me that we CAN easily detect usage of context variables (since we control the accessor methods for them) which is another reason that this move to context variables over params is a good change.

I just merged in the latest from master. If this is starting to come together would you mind kicking off a CI run when convenient?

rjernst commented:
@elasticmachine ok to test

rationull commented:
Merged in the latest from master. Not sure what's going on with the CI failure -- `:server:integTest` passed for me before submitting (and just now I ran it again too). (I ran it on macOS though). The failure doesn't look related.

rjernst commented:
I agree it does not look related, and it also does not reproduce for me. Let's try again.

@elasticmachine test this please

rjernst commented:
Looks like the same failure. @colings86 do you have any ideas? I do not see with that AckIT has to do with scripted metric aggregations.

rationull commented:
Merged in master again and the builds have passed this time.

rjernst commented:
This looks fine from my side now, @colings86 do you want to review?

colings86 commented:
@rationull thanks again for working on this. Are you able to create a PR for removing the deprecation stuff in master (7.0) as well (the TODOs in the code)?

colings86 commented:
Backport pending for 6.x branch, waiting for build on https://github.com/elastic/elasticsearch/pull/32944

rationull commented:
@colings86 absolutely -- I'll make a PR for removing the deprecated stuff this weekend.

colings86 commented:
Awesome, thanks 👍

cbuescher commented:
Removing the "backport pending" label since this has been merged to 6.5 with https://github.com/elastic/elasticsearch/commit/edb577f0316f6ac36e83605d44404291da9750b2
